File: TvEngine3/Filters/sources/TsWriter/source/Technotrend.cpp
/* 
 *	Copyright (C) 2006 Team MediaPortal
 *	http://www.team-mediaportal.com
 *
 *  This Program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2, or (at your option)
 *  any later version.
 *   
 *  This Program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 *  GNU General Public License for more details.
 *   
 *  You should have received a copy of the GNU General Public License
 *  along with GNU Make; see the file COPYING.  If not, write to
 *  the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA. 
 *  http://www.gnu.org/copyleft/gpl.html
 *
 */
#pragma warning(disable : 4995)
#include <windows.h>
#include <commdlg.h>
#include <bdatypes.h>
#include <time.h>
#include <streams.h>
#include <initguid.h>
#include <ks.h>
#include <ksproxy.h>
#include "Technotrend.h"
#include "bdaapi_cimsg.h"


#define LBDG2_NAME_C_TUNER           L"TechnoTrend BDA/DVB-C Tuner"
#define LBDG2_NAME_S_TUNER           L"TechnoTrend BDA/DVB-S Tuner"
#define LBDG2_NAME_T_TUNER           L"TechnoTrend BDA/DVB-T Tuner"
#define LBUDGET3NAME_TUNER           L"TTHybridTV BDA DVBT Tuner"
#define LBUDGET3NAME_TUNER_ANLG      L"TTHybridTV BDA Analog TV Tuner"
#define LUSB2BDA_DVB_NAME_C_TUNER    L"USB 2.0 BDA DVB-C Tuner"
#define LUSB2BDA_DVB_NAME_S_TUNER    L"USB 2.0 BDA DVB-S Tuner"
#define LUSB2BDA_DVB_NAME_T_TUNER    L"USB 2.0 BDA DVB-T Tuner"
#define LUSB2BDA_DVBS_NAME_PIN_TUNER L"Pinnacle PCTV 400e Tuner"


//slot status
#define	CI_SLOT_EMPTY							0
#define	CI_SLOT_MODULE_INSERTED		1
#define	CI_SLOT_MODULE_OK					2
#define CI_SLOT_CA_OK							3
#define	CI_SLOT_DBG_MSG						4
#define	CI_SLOT_UNKNOWN_STATE			0xFF

//SendCIMessage Tags
#define	CI_PSI_COMPLETE					0xC
#define	CI_MODULE_READY					0xD
#define	CI_SWITCH_PRG_REPLY			0xE

extern void LogDebug(const char *fmt, ...) ;

CTechnotrend::CTechnotrend(LPUNKNOWN pUnk, HRESULT *phr)
:CUnknown( NAME ("MpTsTechnoTrend"), pUnk)
{
  m_hBdaApi = INVALID_HANDLE_VALUE;
  m_deviceType=UNKNOWN;
  m_ciStatus=-1;
}
CTechnotrend::~CTechnotrend(void)
{
  if (m_hBdaApi != INVALID_HANDLE_VALUE)
  {
    bdaapiCloseCI(m_hBdaApi);
    bdaapiClose(m_hBdaApi);
  }
}


static void OnSlotStatusCallback(PVOID Context,BYTE nSlot,BYTE nStatus,TYP_SLOT_INFO* csInfo)
{
  CTechnotrend* technoTrend=(CTechnotrend*)Context;
  technoTrend->OnSlotChange( nSlot, nStatus, csInfo);
}
static void OnCaStatusCallback(PVOID Context,BYTE  nSlot,BYTE  nReplyTag,WORD  wStatus)
{
  CTechnotrend* technoTrend=(CTechnotrend*)Context;
  technoTrend->OnCaChange(nSlot,nReplyTag,wStatus);
}

static void OnDisplayString(PVOID Context,BYTE  nSlot,char* pString,WORD  wLength)
{
  LogDebug("TechnoTrend:OnDisplayString slot:%d %s", nSlot,pString);
}
static void OnDisplayMenu(PVOID Context,BYTE  nSlot,WORD  wItems,char* pStringArray,WORD  wLength)
{
  LogDebug("TechnoTrend:OnDisplayMenu slot:%d", nSlot);
}
static void OnDisplayList(PVOID Context,BYTE  nSlot,WORD  wItems,char* pStringArray,WORD  wLength)
{
  LogDebug("TechnoTrend:OnDisplayList slot:%d", nSlot);
}
static void OnSwitchOsdOff(PVOID Context,BYTE  nSlot)
{
  LogDebug("TechnoTrend:CI_OnSwitchOsdOff slot:%d", nSlot);
}
static void OnInputRequest(PVOID Context,BYTE  nSlot,BOOL  bBlindAnswer,BYTE  nExpectedLength, DWORD dwKeyMask)
{
  LogDebug("TechnoTrend:CI_OnInputRequest slot:%d", nSlot);
}
static void OnLscSetDescriptor(PVOID Context,BYTE  nSlot,TYPE_CONNECT_DESCR* pDescriptor)
{
  LogDebug("TechnoTrend:OnLscSetDescriptor slot:%d", nSlot);
}
static void OnLscConnect(PVOID Context,BYTE  nSlot)
{
  LogDebug("TechnoTrend:OnLscConnect slot:%d", nSlot);
}
static void OnLscDisconnect(PVOID Context,BYTE  nSlot)
{
  LogDebug("TechnoTrend:OnLscDisconnect slot:%d", nSlot);
}
static void OnLscSetParams(PVOID Context,BYTE  nSlot,BYTE  BufferSize,BYTE  Timeout10Ms)
{
  LogDebug("TechnoTrend:OnLscSetParams slot:%d", nSlot);
}
static void OnLscEnquireStatus(PVOID Context,BYTE  nSlot)
{
  LogDebug("TechnoTrend:OnLscEnquireStatus slot:%d", nSlot);
}
static void OnLscGetNextBuffer(PVOID Context,BYTE  nSlot,BYTE  PhaseID)
{
  LogDebug("TechnoTrend:OnLscGetNextBuffer slot:%d", nSlot);
}
static void OnLscTransmitBuffer(PVOID Context,BYTE  nSlot,BYTE  PhaseID,BYTE* pData,WORD  nLength)
{
  LogDebug("TechnoTrend:OnLscTransmitBuffer slot:%d", nSlot);
}

STDMETHODIMP CTechnotrend::SetTunerFilter(IBaseFilter* tunerFilter)
{
  FILTER_INFO info;
  if (!SUCCEEDED(tunerFilter->QueryFilterInfo(&info))) return S_OK;
  if (wcscmp(info.achName,LBDG2_NAME_C_TUNER)==0) m_deviceType=BUDGET_2;
  if (wcscmp(info.achName,LBDG2_NAME_S_TUNER)==0) m_deviceType=BUDGET_2;
  if (wcscmp(info.achName,LBDG2_NAME_T_TUNER)==0) m_deviceType=BUDGET_2;
  if (wcscmp(info.achName,LBUDGET3NAME_TUNER)==0) m_deviceType=BUDGET_3;
  if (wcscmp(info.achName,LBUDGET3NAME_TUNER_ANLG)==0) m_deviceType=BUDGET_3;
  if (wcscmp(info.achName,LUSB2BDA_DVB_NAME_C_TUNER)==0) m_deviceType=USB_2;
  if (wcscmp(info.achName,LUSB2BDA_DVB_NAME_S_TUNER)==0) m_deviceType=USB_2;
  if (wcscmp(info.achName,LUSB2BDA_DVB_NAME_T_TUNER)==0) m_deviceType=USB_2;
  if (wcscmp(info.achName,LUSB2BDA_DVBS_NAME_PIN_TUNER)==0) m_deviceType=USB_2_PINNACLE;
  if (m_deviceType==UNKNOWN) return S_OK;
  
  LogDebug("Technotrend: card detected type:%d",m_deviceType);
  UINT deviceId;
  if (!GetDeviceID(tunerFilter, deviceId) )
  {
    LogDebug("Technotrend: unable to determine the device id");
    m_deviceType=UNKNOWN;  
    return S_OK;
  }
  m_hBdaApi = bdaapiOpenHWIdx(m_deviceType, deviceId);
  if (m_hBdaApi == INVALID_HANDLE_VALUE) 
  {
    LogDebug("Technotrend: unable to open the device");
    return S_OK;
  }

  memset(&m_technoTrendStructure,0,sizeof(m_technoTrendStructure));
  m_technoTrendStructure.p01Context=m_technoTrendStructure.p02Context=m_technoTrendStructure.p03Context=m_technoTrendStructure.p04Context=this;
  m_technoTrendStructure.p05Context=m_technoTrendStructure.p06Context=m_technoTrendStructure.p07Context=m_technoTrendStructure.p08Context=this;
  m_technoTrendStructure.p09Context=m_technoTrendStructure.p10Context=m_technoTrendStructure.p11Context=m_technoTrendStructure.p12Context=this;
  m_technoTrendStructure.p13Context=m_technoTrendStructure.p14Context=this;
  m_technoTrendStructure.p01=OnSlotStatusCallback;
  m_technoTrendStructure.p02=OnCaStatusCallback;
  m_technoTrendStructure.p03=OnDisplayString;
  m_technoTrendStructure.p04=OnDisplayMenu;
  m_technoTrendStructure.p05=OnDisplayList;
  m_technoTrendStructure.p06=OnSwitchOsdOff;
  m_technoTrendStructure.p07=OnInputRequest;
  m_technoTrendStructure.p08=OnLscSetDescriptor;
  m_technoTrendStructure.p09=OnLscConnect;
  m_technoTrendStructure.p10=OnLscDisconnect;
  m_technoTrendStructure.p11=OnLscSetParams;
  m_technoTrendStructure.p12=OnLscEnquireStatus;
  m_technoTrendStructure.p13=OnLscGetNextBuffer;
  m_technoTrendStructure.p14=OnLscTransmitBuffer;

  if (!SUCCEEDED(bdaapiOpenCI(m_hBdaApi, m_technoTrendStructure)))
  {
      LogDebug("Technotrend: unable to open the CI");
      m_deviceType=UNKNOWN;  
      return S_OK;
  }
  BYTE v1,v2,v3,v4;
  bdaapiGetDrvVersion(m_hBdaApi,&v1,&v2,&v3,&v4);
  LogDebug("Technotrend: initalized id:%x, driver version:%d.%d.%d.%d",deviceId,v1,v2,v3,v4);
	

  return S_OK;
}

STDMETHODIMP CTechnotrend::IsTechnoTrend( BOOL* yesNo)
{
  *yesNo= (m_hBdaApi != INVALID_HANDLE_VALUE);
  return S_OK;
}

STDMETHODIMP CTechnotrend::IsCamReady( BOOL* yesNo)
{
  *yesNo=FALSE;
  if (m_slotStatus==CI_SLOT_CA_OK || m_slotStatus==CI_SLOT_MODULE_OK)
  {
    *yesNo=TRUE;
  }
  return S_OK;
}

STDMETHODIMP CTechnotrend::SetAntennaPower( BOOL onOff)
{
  if (m_deviceType==USB_2_PINNACLE || m_deviceType==USB_2)
  {
    HRESULT hr=bdaapiSetDVBTAntPwr(m_hBdaApi,onOff);
    LogDebug("Technotrend: enable antenna power:%d %x",onOff,hr);
  }
  return S_OK;
}

STDMETHODIMP CTechnotrend::SetDisEqc(int diseqcType, int hiband, int vertical)
{
	LogDebug("Technotrend: SetDisEqc antenna :%d hiband:%d vertical:%d",diseqcType,hiband, vertical);
  int antennaNr=1;
  int option=0;
  switch (diseqcType)
  {
    case 0:
    case 1://simple A
      antennaNr = 1;
      break;
    case 2://simple B
      antennaNr = 2;
      break;
    case 3://Level 1 A/A
      antennaNr = 1;
      break;
    case 4://Level 1 B/A
      antennaNr = 2;
      break;
    case 5://Level 1 A/B
      antennaNr = 3;
      break;
    case 6://Level 1 B/B
      antennaNr = 4;
      break;
  }
  ULONG diseqc = 0xE01038F0;
  if (hiband!=0)              // high band
    diseqc |= 0x00000001;

  if (vertical==0)            // horizontal
    diseqc |= 0x00000002;

  diseqc |=  (byte)((antennaNr - 1) << 2);

  Polarisation polarity;
  if (vertical)
    polarity = BDA_POLARISATION_LINEAR_V;
  else
    polarity = BDA_POLARISATION_LINEAR_H;

  BYTE data[4];
  data[0]=(BYTE)((diseqc >> 24) & 0xff);
  data[1]=(BYTE)((diseqc >> 16) & 0xff);
  data[2]=(BYTE)((diseqc >> 8) & 0xff);
  data[3]=(BYTE)((diseqc ) & 0xff);

  HRESULT hr=bdaapiSetDiSEqCMsg(m_hBdaApi,&data[0],4,0,0,polarity);
  LogDebug("TechnoTrend:SetDiseqc:%x %x", diseqc,hr);
	bdaapiCIGetSlotStatus(m_hBdaApi,0);
	
  return S_OK;
}

STDMETHODIMP CTechnotrend::DescrambleService( int serviceId,BOOL* succeeded)
{
  HRESULT hr;
  *succeeded=FALSE;
  BOOL enabled=FALSE;
  m_ciStatus=-1;
	bdaapiCIGetSlotStatus(m_hBdaApi,0);
	LogDebug("TechnoTrend: DescrambleService:0x%x (%d) (%d)",serviceId,serviceId,m_slotStatus);
	if (m_slotStatus==CI_SLOT_CA_OK || m_slotStatus==CI_SLOT_MODULE_OK )
  {
    hr = bdaapiCIReadPSIFastDrvDemux(m_hBdaApi, (WORD)serviceId);
    if (hr==RET_SUCCESS)
    {
      if (m_ciStatus==1)
      {
        *succeeded=TRUE;
        LogDebug("TechnoTrend: service decoded:%x %d",hr,m_ciStatus);
    
      }
      else
      {
        LogDebug("TechnoTrend: service decoded:%x %d",hr,m_ciStatus);
      }
    }
    else
    {
      LogDebug("TechnoTrend: service not decoded:%x",hr);
    }    
  }
	else if (m_slotStatus==CI_SLOT_UNKNOWN_STATE || m_slotStatus==CI_SLOT_EMPTY)
	{
		//no CAM inserted
      LogDebug("TechnoTrend: no cam detected:%d",m_slotStatus);
    *succeeded=TRUE;
	}
  
  return S_OK;
}

bool CTechnotrend::GetDeviceID(IBaseFilter* tunerFilter, UINT& deviceId)
{
  bool success=false;
  IEnumPins* pEnumPins;
  tunerFilter->EnumPins(&pEnumPins);
  ULONG fetched;
  IPin* pins[2];
  while (SUCCEEDED( pEnumPins->Next(1,&pins[0],&fetched)))
  {
    if (fetched!=1) break;
    PIN_DIRECTION pinDirection;
    pins[0]->QueryDirection(&pinDirection);
    if (pinDirection!= PINDIR_OUTPUT) 
    {
      pins[0]->Release();
      continue;
    }

    IKsPin* pKsPin;
    if (SUCCEEDED( pins[0]->QueryInterface(IID_IKsPin, (void **)&pKsPin)))
    {    
      KSMULTIPLE_ITEM *pmi;
      HRESULT hr=pKsPin->KsQueryMediums(&pmi);
      pKsPin->Release();
      if (SUCCEEDED(hr))
      {
        // Use pointer arithmetic to reference the first medium structure.
        REGPINMEDIUM *pTemp = (REGPINMEDIUM*)(pmi + 1);
        for (ULONG i = 0; i < pmi->Count; i++, pTemp++) 
        {
          success=true;
          deviceId=pTemp->dw1;
          break;
        }
        CoTaskMemFree(pmi);
      }
    }
    pins[0]->Release();
  }
  pEnumPins->Release();
  return success;

}


void CTechnotrend::OnCaChange(BYTE  nSlot,BYTE  nReplyTag,WORD  wStatus)
{
	LogDebug("$ OnCaChange slot:%d reply:%d status:%d",nSlot,nReplyTag,wStatus);
	switch(nReplyTag)	
	{
		case CI_PSI_COMPLETE:
			LogDebug("$ CI: ### Number of programs : %04d",wStatus);
		break;

		case CI_MODULE_READY:
			LogDebug("$ CI: CI_MODULE_READY in OnCAStatus not supported");
		break;
		case CI_SWITCH_PRG_REPLY:
		{
			switch(wStatus)
			{
				case ERR_INVALID_DATA:
					LogDebug("$ CI: ERROR::SetProgram failed !!! (invalid PNR)");
				break;
				case ERR_NO_CA_RESOURCE:
					LogDebug("$ CI: ERROR::SetProgram failed !!! (no CA resource available)");
				break;
				case ERR_NONE:
					LogDebug("$ CI:    SetProgram OK");
					m_ciStatus=1;
				break;
				default:
				break;
			}
		}
		break;
		default:
		break;
	}
}

void CTechnotrend::OnSlotChange(BYTE nSlot,BYTE nStatus,TYP_SLOT_INFO* csInfo)
{
  if (nStatus==0) LogDebug("Technotrend: slot:%d empty",nSlot);  
  else if (nStatus==1) LogDebug("Technotrend: slot:%d module inserted",nSlot);
  else if (nStatus==2) LogDebug("Technotrend: slot:%d module ok",nSlot);
  else if (nStatus==3) LogDebug("Technotrend: slot:%d ca ok",nSlot);
  else if (nStatus==4) LogDebug("Technotrend: slot:%d dbg msg",nSlot);
  else  LogDebug("Technotrend: slot:%d unknown state:%x",nSlot,nStatus);
  m_slotStatus=nStatus;
	
  if (csInfo!=NULL)
  {
    LogDebug("Technotrend:    CI status:%d ",csInfo->nStatus);
    if (csInfo->pMenuTitleString!=NULL)
      LogDebug("Technotrend:    CI text  :%s ",csInfo->pMenuTitleString);
    for (int i=0; i < csInfo->wNoOfCaSystemIDs;++i)
    {
      LogDebug("Technotrend:      ca system id  :%x ",csInfo->pCaSystemIDs[i]);
    }
  }
}

File: TV-Server 1.0/TVLibrary/TVLibrary/Implementations/DVB/ConditionalAccess/TechnoTrend/TechnoTrend.cs
/* 
 *	Copyright (C) 2005-2008 Team MediaPortal
 *	http://www.team-mediaportal.com
 *
 *  This Program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2, or (at your option)
 *  any later version.
 *   
 *  This Program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 *  GNU General Public License for more details.
 *   
 *  You should have received a copy of the GNU General Public License
 *  along with GNU Make; see the file COPYING.  If not, write to
 *  the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA. 
 *  http://www.gnu.org/copyleft/gpl.html
 *
 */
using System;
using System.Collections;
using System.Collections.Generic;
using System.Runtime.InteropServices;
using DirectShowLib;
using DirectShowLib.BDA;
using System.Windows.Forms;
using TvLibrary.Channels;
using TvLibrary.Interfaces.Analyzer;

namespace TvLibrary.Implementations.DVB
{
  public class TechnoTrend : IDisposable, IDiSEqCController
  {
    ITechnoTrend _technoTrendInterface = null;
    IntPtr ptrPmt;
    IntPtr _ptrDataInstance;
    DVBSChannel _previousChannel = null;
    /// <summary>
    /// Initializes a new instance of the <see cref="TechnoTrend"/> class.
    /// </summary>
    /// <param name="tunerFilter">The tuner filter.</param>
    /// <param name="analyzerFilter">The analyzer filter.</param>
    public TechnoTrend(IBaseFilter tunerFilter, IBaseFilter analyzerFilter)
    {
      _technoTrendInterface = analyzerFilter as ITechnoTrend;
      _technoTrendInterface.SetTunerFilter(tunerFilter);
      ptrPmt = Marshal.AllocCoTaskMem(1024);
      _ptrDataInstance = Marshal.AllocCoTaskMem(1024);
    }

    /// <summary>
    /// Performs application-defined tasks associated with freeing, releasing, or resetting unmanaged resources.
    /// </summary>
    public void Dispose()
    {
      _technoTrendInterface = null;
      Marshal.FreeCoTaskMem(ptrPmt);
      Marshal.FreeCoTaskMem(_ptrDataInstance);
    }

    /// <summary>
    /// Determines whether cam is ready.
    /// </summary>
    /// <returns>
    /// 	<c>true</c> if [is cam ready]; otherwise, <c>false</c>.
    /// </returns>
    public bool IsCamReady()
    {
      if (_technoTrendInterface == null) return false;
      bool yesNo = false;
      _technoTrendInterface.IsCamReady(ref yesNo);
      return yesNo;
    }

    /// <summary>
    /// Gets a value indicating whether this instance is techno trend.
    /// </summary>
    /// <value>
    /// 	<c>true</c> if this instance is techno trend; otherwise, <c>false</c>.
    /// </value>
    public bool IsTechnoTrend
    {
      get
      {
        if (_technoTrendInterface == null) return false;
        bool yesNo = false;
        _technoTrendInterface.IsTechnoTrend(ref yesNo);
        return yesNo;
      }
    }

    /// <summary>
    /// Sends the PMT.
    /// </summary>
    /// <param name="pmt">The PMT.</param>
    /// <param name="PMTlength">The PM tlength.</param>
    /// <returns></returns>
    public bool SendPMT(byte[] pmt, int PMTlength)
    {
      if (_technoTrendInterface == null) return true;
      bool succeeded = false;
      for (int i = 0; i < PMTlength; ++i)
      {
        Marshal.WriteByte(ptrPmt, i, pmt[i]);
      }
      _technoTrendInterface.DescrambleService(ptrPmt, (short)PMTlength, ref succeeded);
      return succeeded;
    }

    /// <summary>
    /// Instructs the technotrend card to descramble all programs mentioned in subChannels.
    /// </summary>
    /// <param name="subChannels">The sub channels.</param>
    /// <returns></returns>
    public bool DescrambleMultiple(Dictionary<int, ConditionalAccessContext> subChannels)
    {
      if (_technoTrendInterface == null) return true;
      List<ConditionalAccessContext> filteredChannels = new List<ConditionalAccessContext>();
      bool succeeded = true;
      Dictionary<int, ConditionalAccessContext>.Enumerator en = subChannels.GetEnumerator();
      while (en.MoveNext())
      {
        bool exists = false;
        ConditionalAccessContext context = en.Current.Value;
        foreach (ConditionalAccessContext c in filteredChannels)
        {
          if (c.Channel.Equals(context.Channel)) exists = true;
        }
        if (!exists)
        {
          filteredChannels.Add(context);
        }
      }

      for (int i = 0; i < filteredChannels.Count; ++i)
      {
        ConditionalAccessContext context = filteredChannels[i];
        Marshal.WriteInt16(ptrPmt, 2 * i, (short)context.ServiceId);
      }
      _technoTrendInterface.DescrambleMultiple(ptrPmt, (short)filteredChannels.Count, ref succeeded);
      return succeeded;
    }

    /// <summary>
    /// Sends the diseq command.
    /// </summary>
    /// <param name="channel">The channel.</param>
    /// <param name="parameters">The scanparameters.</param>
    public void SendDiseqCommand(ScanParameters parameters, DVBSChannel channel)
    {
      if (_technoTrendInterface == null) return;
      if (_previousChannel != null)
      {
        if (_previousChannel.Frequency == channel.Frequency &&
            _previousChannel.DisEqc == channel.DisEqc &&
            _previousChannel.Polarisation == channel.Polarisation)
        {
          Log.Log.WriteFile("Technotrend: already tuned to diseqc:{0}, frequency:{1}, polarisation:{2}",channel.DisEqc, channel.Frequency, channel.Polarisation);
          return;
        }
      }
      _previousChannel = channel;
      int antennaNr = BandTypeConverter.GetAntennaNr(channel);
      //"01,02,03,04,05,06,07,08,09,0a,0b,cc,cc,cc,cc,cc,cc,cc,cc,cc,cc,cc,cc,cc,cc,"	
      Marshal.WriteByte(_ptrDataInstance, 0, 0xE0);//diseqc command 1. uFraming=0xe0
      Marshal.WriteByte(_ptrDataInstance, 1, 0x10);//diseqc command 1. uAddress=0x10
      Marshal.WriteByte(_ptrDataInstance, 2, 0x38);//diseqc command 1. uCommand=0x38
      //bit 0	(1)	: 0=low band, 1 = hi band
      //bit 1 (2) : 0=vertical, 1 = horizontal
      //bit 3 (4) : 0=satellite position A, 1=satellite position B
      //bit 4 (8) : 0=switch option A, 1=switch option  B
      // LNB    option  position
      // 1        A         A
      // 2        A         B
      // 3        B         A
      // 4        B         B
      bool hiBand = BandTypeConverter.IsHiBand(channel,parameters);
      Log.Log.WriteFile("TechnoTrend SendDiseqcCommand() diseqc:{0}, antenna:{1} frequency:{2}, polarisation:{3} hiband:{4}", channel.DisEqc, antennaNr, channel.Frequency, channel.Polarisation, hiBand);
      bool isHorizontal = ((channel.Polarisation == Polarisation.LinearH) || (channel.Polarisation == Polarisation.CircularL));
      byte cmd = 0xf0;
      cmd |= (byte)(hiBand ? 1 : 0);
      cmd |= (byte)((isHorizontal) ? 2 : 0);
      cmd |= (byte)((antennaNr - 1) << 2);
      Marshal.WriteByte(_ptrDataInstance, 3, cmd);
      _technoTrendInterface.SetDisEqc(_ptrDataInstance, 4, 1, 0, (short)channel.Polarisation);
    }

    /// <summary>
    /// Determines whether [is cam present].
    /// </summary>
    /// <returns>
    /// 	<c>true</c> if [is cam present]; otherwise, <c>false</c>.
    /// </returns>
    public bool IsCamPresent()
    {
      if (_technoTrendInterface == null) return false;
      bool yesNo = false;
      _technoTrendInterface.IsCamPresent(ref yesNo);
      return yesNo;
    }

    #region IDiSEqCController Members

    /// <summary>
    /// Sends the DiSEqC command.
    /// </summary>
    /// <param name="diSEqC">The DiSEqC command.</param>
    /// <returns>true if succeeded, otherwise false</returns>
    public bool SendDiSEqCCommand(byte[] diSEqC)
    {
      if (_technoTrendInterface == null) return false;
      for (int i = 0; i < diSEqC.Length; ++i)
        Marshal.WriteByte(_ptrDataInstance, i, diSEqC[i]);
      Polarisation pol = Polarisation.LinearV;
      if (_previousChannel != null)
        pol = _previousChannel.Polarisation;
      _technoTrendInterface.SetDisEqc(_ptrDataInstance, (byte)diSEqC.Length, 1, 0, (short)pol);
      return true;
    }

    /// <summary>
    /// gets the diseqc reply
    /// </summary>
    /// <param name="reply">The reply.</param>
    /// <returns>true if succeeded, otherwise false</returns>
    public bool ReadDiSEqCCommand(out byte[] reply)
    {
      reply = null;
      return false;
    }

    #endregion
  }
}
